base/math.jl

```diff
@@ -536,38 +536,56 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+	#= Keep specialized two argument version, because it is faster
+	than generic varargs version below =#
+	ay, ax = minmax(abs(x), abs(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
     rr = ax * sqrt(1 + r * r)
 
-    # Use type of rr to make sure that return type is the same for
-    # all branches
-    if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
-    else
-        return rr
-    end
+	if isnan(r)
+		if isinf(x) || isinf(y)
+			return oftype(rr, Inf)
+		else
+			return oftype(rr, NaN)
+		end
+   	else
+   		return rr
+   	end
 end
-
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	if isnan(s)
+		return any(isinf, x) ? oftype(s, Inf) : s # IEEE 754
```

Comment:
This is the wrong return type.  If `x` is dimensionful, then hypot should return something with the same units as `x`, but this returns a value with the units of `x^2`.

Just do `oftype(sqrt(s), Inf) : sqrt(s)` — the extra `sqrt` call is harmless since we don't need to optimize the NaN case, and the `oftype` branch will probably optimize away the call anyway.

Reply:
We wanted to avoid the extra `sqrt` call. In a previous commit we had `r = sqrt(s)` at the top, and then used `r` to decide the return type in all branches. I just benchmarked this and it is slower (seems the compiler is not able to recognize that we only need the type of `r`, not its value). I also benchmarked a version of the code where I used `sqrt(s)` instead of `r` wherever it was needed, but still the compiler is not optimizing the call away. Any suggestions?

Reply:
How can I write a test to with a dimensionful `x`?

Reply:
For testing dimensionful `x`, see the tests with `Furlongs` in `tests/math.jl`.  There actually are a few tests for `hypot` already.

Reply:
Ah, that's what the "Furlong" was. It was only testing the two-arg version. I added some var args tests. Thanks